src/transport/SecureSessionMgr.cpp

```diff
@@ -166,18 +175,24 @@ CHIP_ERROR SecureSessionMgr::SendMessage(SecureSessionHandle session, PayloadHea
         SuccessOrExit(err);
     }
 
+#if !CHIP_SYSTEM_CONFIG_USE_LWIP1
     // The start of buffer points to the beginning of the encrypted header, and the length of buffer
     // contains both the encrypted header and encrypted data.
     // Locally store the start and length of the retained buffer after accounting for the size of packet header.
     headerSize = packetHeader.EncodeSizeBytes();
 
     msgStart = static_cast<uint8_t *>(msgBuf->Start() - headerSize);
     msgLen   = static_cast<uint16_t>(msgBuf->DataLength() + headerSize);
+#endif
 
     // Retain the packet buffer in case it's needed for retransmissions.
     if (bufferRetainSlot != nullptr)
     {
+#if CHIP_SYSTEM_CONFIG_USE_LWIP1        
+        encryptedMsg        = msgBuf.CloneData(MessagePacketBuffer::kMaxFooterSize);
```

Comment:
So if at this point we encoded the packetHeader into this buffer, such that we maintain the invariant that our retained buffer has a packetHeader encoded into it, wouldn't that remove a lot of the other complexity being added here?  That seems like it would let us remove all the ifdef bits above and have a single code flow...

If we also moved the setting of msgStart/msgLen into the place where we set `encryptedMsg`, we could get rid of the ifdef around the following SetStart/SetDataLength calls.  So the only ifdef we would need is in this one spot.

Reply:
That is my initial trying, but it also did not work, I think the assumption of this solution is Transport layer will send the encrypted packet as is without any magic, but it seems this is not the truth.

+ @pan-apple  for comments.

Reply:
ping @pan-apple

Reply:
@yufengwangca Which case did not work?  The case when things were being cloned, or the case when things were not being cloned?  Do you still have that initial attempt that did not work?  I'd really like to see the code involved....

Reply:
@bzbarsky-apple Actually, I found we can solve this issue without expensive "deep copy" methods we discussed above. Lets follow-up this issue in https://github.com/project-chip/connectedhomeip/pull/5309